pstoreds/addr_book_gc.go

```diff
@@ -0,0 +1,387 @@
+	}
+	if ab.opts.GCInitialDelay < 0 {
+		return nil, fmt.Errorf("negative GC initial delay provided: %s", ab.opts.GCInitialDelay)
+	}
+	if ab.opts.GCLookaheadInterval > 0 && ab.opts.GCLookaheadInterval < ab.opts.GCPurgeInterval {
+		return nil, fmt.Errorf("lookahead interval must be larger than purge interval, respectively: %s, %s",
+			ab.opts.GCLookaheadInterval, ab.opts.GCPurgeInterval)
+	}
+
+	lookaheadEnabled := ab.opts.GCLookaheadInterval > 0
+	gc := &dsAddrBookGc{
+		ctx:              ctx,
+		ab:               ab,
+		running:          make(chan struct{}, 1),
+		lookaheadEnabled: lookaheadEnabled,
+	}
+
+	if lookaheadEnabled {
+		gc.purgeFunc = gc.purgeLookahead
+	} else {
+		gc.purgeFunc = gc.purgeStore
+	}
+
+	// do not start GC timers if purge is disabled; this GC can only be triggered manually.
+	if ab.opts.GCPurgeInterval > 0 {
+		go gc.background()
+	}
+
+	return gc, nil
+}
+
+// gc prunes expired addresses from the datastore at regular intervals. It should be spawned as a goroutine.
+func (gc *dsAddrBookGc) background() {
+	gc.ab.done.Add(1)
+	defer gc.ab.done.Done()
+
+	select {
+	case <-time.After(gc.ab.opts.GCInitialDelay):
+	case <-gc.ab.ctx.Done():
+		// yield if we have been cancelled/closed before the delay elapses.
+		return
+	}
+
+	purgeTimer := time.NewTicker(gc.ab.opts.GCPurgeInterval)
+	defer purgeTimer.Stop()
+
+	var lookaheadCh <-chan time.Time
+	if gc.lookaheadEnabled {
+		lookaheadTimer := time.NewTicker(gc.ab.opts.GCLookaheadInterval)
+		lookaheadCh = lookaheadTimer.C
+		gc.populateLookahead() // do a lookahead now
+		defer lookaheadTimer.Stop()
+	}
+
+	for {
+		select {
+		case <-purgeTimer.C:
+			gc.purgeFunc()
+
+		case <-lookaheadCh:
+			// will never trigger if lookahead is disabled (nil Duration).
+			gc.populateLookahead()
+
+		case <-gc.ctx.Done():
+			return
+		}
+	}
+}
+
+// purgeCycle runs a single GC purge cycle. It operates within the lookahead window if lookahead is enabled; else it
+// visits all entries in the datastore, deleting the addresses that have expired.
+func (gc *dsAddrBookGc) purgeLookahead() {
+	select {
+	case gc.running <- struct{}{}:
+		defer func() { <-gc.running }()
+	default:
+		// yield if lookahead is running.
+		return
+	}
+
+	var id peer.ID
+	record := &addrsRecord{AddrBookRecord: &pb.AddrBookRecord{}} // empty record to reuse and avoid allocs.
+	batch, err := newCyclicBatch(gc.ab.ds, defaultOpsPerCyclicBatch)
+	if err != nil {
+		log.Warningf("failed while creating batch to purge GC entries: %v", err)
+	}
+
+	// This function drops an unparseable GC entry; this is for safety. It is an escape hatch in case
+	// we modify the format of keys going forward. If a user runs a new version against an old DB,
+	// if we don't clean up unparseable entries we'll end up accumulating garbage.
+	dropInError := func(key ds.Key, err error, msg string) {
+		if err != nil {
+			log.Warningf("failed while %s with GC key: %v, err: %v", msg, key, err)
```

Comment:
I would add info that we are removing the record.

Reply:
+1. Done.